src/twisted/protocols/_smb/ntlm.py

```diff
@@ -0,0 +1,380 @@
+# -*- test-case-name: twisted.protocols._smb.tests -*-
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+"""Implement the NT Lan Manager (NTLMv2) challenge/response authentication
+protocol
+ """
+
+from zope.interface import implementer
+
+import struct
+import time
+import socket
+import hmac
+import hashlib
+
+from twisted.protocols._smb import base
+
+import twisted.cred.credentials
+from twisted.python.randbytes import secureRandom
+from twisted.logger import Logger
+
+log = Logger()
+
+NTLM_MESSAGES = ['invalid', 'negotiate', 'challenge', 'auth']
+FLAGS = {
+    'NegotiateUnicode': 0x00000001,
+    'NegotiateOEM': 0x00000002,
+    'RequestTarget': 0x00000004,
+    'Unknown9': 0x00000008,
+    'NegotiateSign': 0x00000010,
+    'NegotiateSeal': 0x00000020,
+    'NegotiateDatagram': 0x00000040,
+    'NegotiateLanManagerKey': 0x00000080,
+    'Unknown8': 0x00000100,
+    'NegotiateNTLM': 0x00000200,
+    'NegotiateNTOnly': 0x00000400,
+    'Anonymous': 0x00000800,
+    'NegotiateOemDomainSupplied': 0x00001000,
+    'NegotiateOemWorkstationSupplied': 0x00002000,
+    'Unknown6': 0x00004000,
+    'NegotiateAlwaysSign': 0x00008000,
+    'TargetTypeDomain': 0x00010000,
+    'TargetTypeServer': 0x00020000,
+    'TargetTypeShare': 0x00040000,
+    'NegotiateExtendedSecurity': 0x00080000,
+    'NegotiateIdentify': 0x00100000,
+    'Unknown5': 0x00200000,
+    'RequestNonNTSessionKey': 0x00400000,
+    'NegotiateTargetInfo': 0x00800000,
+    'Unknown4': 0x01000000,
+    'NegotiateVersion': 0x02000000,
+    'Unknown3': 0x04000000,
+    'Unknown2': 0x08000000,
+    'Unknown1': 0x10000000,
+    'Negotiate128': 0x20000000,
+    'NegotiateKeyExchange': 0x40000000,
+    'Negotiate56': 0x80000000
+}
+
+DEFAULT_FLAGS = {
+    "NegotiateUnicode", "NegotiateSign", "RequestTarget", "NegotiateNTLM",
+    "NegotiateAlwaysSign", "NegotiateExtendedSecurity", "NegotiateTargetInfo",
+    "NegotiateVersion", "Negotiate128", "NegotiateKeyExchange", "Negotiate56"
+}
+
+
+
+def flags2set(flags):
+    """
+    convert C-style flags to Python set
+
+    @param flags: the flags
+    @type flags: L{int}
+
+    @rtype: L{set} of L{str}
+    """
+    r = set()
+    for k, v in FLAGS.items():
+        if v | flags > 0:
+            r.add(k)
+    return r
+
+
+
+def set2flags(s):
+    """
+    convert set to C-style flags
+
+    @rtype: L{int}
+
+    @type s: L{set} of L{str}
+    """
+    flags = 0
+    for i in s:
+        flags |= FLAGS[i]
+    return flags
+
```

Comment:
I think you might get a lot of what you want here out of `IntFlag`

Reply:
Correct. `IntFlag` was introduced in 3 6